exploratory/app.R

```diff
@@ -1,62 +1,60 @@
-options("teal.ggplot2_args" = teal.widgets::ggplot2_args(labs = list(caption = "NEST PROJECT")))
-
 library(teal.modules.general)
-library(scda)
-library(scda.2022)
-library(dplyr)
-library(tidyr)
-library(nestcolor)
-# optional libraries
-library(broom)
-library(colourpicker)
-library(ggExtra)
-library(ggpmisc)
-library(ggpp)
-library(goftest)
-library(gridExtra)
-library(htmlwidgets)
-library(jsonlite)
-library(lattice)
-library(MASS)
-library(rlang)
-library(rtables)
-library(sparkline)
-
-options(shiny.useragg = FALSE)
+options(
+  shiny.useragg = FALSE,
+  teal.ggplot2_args = teal.widgets::ggplot2_args(labs = list(caption = "NEST PROJECT"))
+)
 
 nest_logo <- "https://raw.githubusercontent.com/insightsengineering/hex-stickers/main/PNG/nest.png"
 
-# code>
-ADSL <- synthetic_cdisc_data("latest")$adsl
-ADRS <- synthetic_cdisc_data("latest")$adrs
-ADLB <- synthetic_cdisc_data("latest")$adlb
+## Data reproducible code ----
+data <- teal_data()
+data <- within(data, {
+  library(scda)
+  library(scda.2022)
+  library(dplyr)
+  library(tidyr)
+  library(ggExtra)
+  library(ggpmisc)
+  library(ggpp)
+  library(goftest)
+  library(gridExtra)
+  library(htmlwidgets)
+  library(jsonlite)
+  library(lattice)
+  library(MASS)
+  library(rlang)
+  library(rtables)
+  library(nestcolor)
+  library(broom)
+  library(colourpicker)
+  library(sparkline)
 
-ADLBPCA <- ADLB %>%
-  dplyr::select(USUBJID, STUDYID, SEX, ARMCD, AVAL, AVISIT, PARAMCD) %>%
-  tidyr::pivot_wider(
-    values_from = "AVAL",
-    names_from = c("PARAMCD", "AVISIT"),
-    names_sep = " - "
-  )
 
-adsl <- cdisc_dataset("ADSL", ADSL, code = "ADSL <- synthetic_cdisc_data(\"latest\")$adsl")
-adrs <- cdisc_dataset("ADRS", ADRS, code = "ADRS <- synthetic_cdisc_data(\"latest\")$adrs")
-adlb <- cdisc_dataset("ADLB", ADLB, code = "ADLB <- synthetic_cdisc_data(\"latest\")$adlb")
-adlbpca <- cdisc_dataset(
-  "ADLBPCA",
-  ADLBPCA,
-  code = 'ADLBPCA <- ADLB %>%
+  ADSL <- synthetic_cdisc_data("latest")$adsl
+  ADRS <- synthetic_cdisc_data("latest")$adrs
+  ADLB <- synthetic_cdisc_data("latest")$adlb
+
+  ADLBPCA <- ADLB %>%
     dplyr::select(USUBJID, STUDYID, SEX, ARMCD, AVAL, AVISIT, PARAMCD) %>%
     tidyr::pivot_wider(
       values_from = "AVAL",
       names_from = c("PARAMCD", "AVISIT"),
       names_sep = " - "
-    )',
-  keys = c("STUDYID", "USUBJID"),
-  label = "ADLB reshaped",
-  vars = list(ADLB = adlb)
-)
-# <code
+    )
+})
+
+datanames <- c("ADSL", "ADRS", "ADLB", "ADLBPCA")
+datanames(data) <- datanames
+
+join_keys(data) <- default_cdisc_join_keys[datanames]
+join_keys(data)[["ADLBPCA"]][["ADLBPCA"]] <- c("USUBJID", "STUDYID", "PARAMCD", "AVISIT")
+join_keys(data)[["ADLBPCA"]][["ADSL"]] <- c("USUBJID", "STUDYID")
```

Comment:
There is a nice feature in `join_keys` which allows to change name of the dataset. Here "ADLBPCA" is a dataset obtained from ADLB - so we can reuse ADLB keys. I don't suggest this change here but I'd like to point out that is possible and you might prefer this in other places maybe. FYI to all

@vedhav @kartikeyakirar @chlebowa @m7pr @averissimo @donyunardi @pawelru 

```suggestion
join_keys(data) <- default_cdisc_join_keys[datanames]
adlbpca_keys <- default_cdisc_join_keys["ADLB"]
names(adlbpca_keys)[1] <- "ADLBPCA"
join_keys(data) <- c(join_keys(data), adlbpca_keys)
```

Reply:
I think some implicit mappings might go missing for `ADLBPCA` unless we explicitly connect `ADLBPCA` with `ADLB` because the `ADLB` is only connected with `ADSL` as a child and only this relationship is retained when we copy like this. We also need all the implicit connections from `ADLB` available for `ADLBPCA`

```r
join_keys(data) <- default_cdisc_join_keys[datanames]
adlbpca_keys <- default_cdisc_join_keys["ADLB"]
names(adlbpca_keys)[1] <- "ADLBPCA"
join_keys(data) <- c(join_keys(data), adlbpca_keys)
join_keys(data)
# ADLBPCA: [STUDYID, USUBJID, PARAMCD, AVISIT]
#   <-> ADSL: [STUDYID, USUBJID] 

join_keys(data)["ADLBPCA", "ADLB"] <- c("USUBJID", "STUDYID")
# ADLBPCA: [STUDYID, USUBJID, PARAMCD, AVISIT]
#   <-> ADSL: [STUDYID, USUBJID]
#   <-> ADLB: [USUBJID, STUDYID]

# For more accurate relationship mapping, although not needed
parents(data) <- c(parents(data), list(ADLBPCA = "ADLB"))
# ADLBPCA: [STUDYID, USUBJID, PARAMCD, AVISIT]
#   <-> ADSL: [STUDYID, USUBJID]
#   --> ADLB: [USUBJID, STUDYID]
```

Reply:
It makes me thinking if we should conclude implicit keys on grandchildren also @averissimo

Reply:
These examples will confuse me as a new user who does not know the mechanics of join_keys. We should encourage people to create their own keys from scratch, chances are their keys have nothing to do with our default set of keys.

And I think that parent mappings are not copied when we do a `c` merge of join keys.

```r
jk_adlbpca <- join_keys(
  join_key("ADLBPCA", "ADLBPCA", c("STUDYID", "USUBJID", "PARAMCD", "AVISIT")),
  join_key("ADLBPCA", "ADLB", c("STUDYID", "USUBJID", "PARAMCD", "AVISIT"))
)
parents(jk_adlbpca) <- list(ADLBPCA = "ADLB")

jk <- default_cdisc_join_keys["ADLB"]
jk <- c(jk, jk_adlbpca)
parents(jk)
# $ADLB
# [1] "ADSL"
parents(jk) <- c(parents(jk), list("ADLBPCA" = "ADLB"))
parents(jk)
# $ADLB
# [1] "ADSL"
#
# $ADLBPCA
# [1] "ADLB"
```